### PR TITLE
Bug fix/do not reclaim shard ownership.

### PR DESCRIPTION
**Pull request by neunhoef**
During a move-shard job which moves the leader there is a
situation in which the old owner of a shard can reclaim ownership
(after having resigned already), with a small race where it allows to
write documents only locally, but then continue the move-shard to a
server without those documents. An additional bug in the MoveShard
Supervision job would then leave the shard in a bad configuration
with a resigned leader permanently in charge.

### Scope & Purpose

*(Can you describe what functional change your PR is trying to effect?)*

- [ ] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [X] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [ ] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [ ] The behavior change can be verified via automatic tests

#### Related Information

*(Please reference tickets / specification etc )*

- [ ] There is a GitHub Issue reported by a Community User: 
- [ ] There is an internal planning ticket: 
- [ ] There is a *JIRA Ticket number* (In case a customer was affected / involved): 
- [X] There is a design document: https://github.com/arangodb/documents/blob/master/DesignDocuments/99_UNSORTED/Supervision.md

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [ ] Added **Regression Tests** (Only for bug-fixes) 
- [X] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
   - Did you add tests for a new *RestHandler* subclass ?
   - Did you add new mocks of underlying code layers to be able to verify your functionality ? 
   - ...
- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
- [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [X] There are tests in an external testing repository (i.e. node-resilience tests, chaos tests): arangosync
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11539/

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11546/

**Comment by neunhoef** [comment]
This run should fix the failures of the previous one.

File: arangod/Cluster/Maintenance.cpp [hunk 1]
```diff
@@ -261,40 +271,28 @@ void handlePlanShard(VPackSlice const& cprops, VPackSlice const& ldb,
...
-      // a leader restart.
-      shouldBeLeading = true;
-      shouldResign = false;
-    }
-    if ((leading != shouldBeLeading && !shouldResign) || !leaderTouched) {
+    /*
+    LOG_DEVEL_IF(dbname=="foo") << "Col: " << colname << " shard: " << shname;
+    LOG_DEVEL_IF(dbname=="foo") << std::boolalpha << "leading: " << leading << " should: " << shouldBeLeading << " resign: " << shouldResign;
+    LOG_DEVEL_IF(dbname=="foo") << "New leader: " << leaderId << " currentLeader: " << localLeader;
+    */
```
Comment (mchacki) [inline arangod/Cluster/Maintenance.cpp:278]:
```suggestion
```
Sorry I left this in there myself.

File: arangod/Cluster/Maintenance.cpp [hunk 2]
```diff
@@ -261,40 +271,28 @@ void handlePlanShard(VPackSlice const& cprops, VPackSlice const& ldb,
...
-                {OLD_CURRENT_COUNTER, std::to_string(feature.getCurrentCounter())}},
-                LEADER_PRIORITY));
+          std::map<std::string, std::string>{
+              {NAME, TAKEOVER_SHARD_LEADERSHIP},
+              {DATABASE, dbname},
+              {COLLECTION, colname},
+              {SHARD, shname},
+              {THE_LEADER, std::string()},
+              {LOCAL_LEADER, std::string(localLeader)},
+              {OLD_CURRENT_COUNTER, "0"}},   // legacy, no longer used
```
Comment (mchacki) [inline arangod/Cluster/Maintenance.cpp:294]:
```suggestion
              {OLD_CURRENT_COUNTER, std::to_string(feature.getCurrentCounter())}}, 
```
i am not too deep into the behaviors here, however in this version the Counter was filled with a sensible value.
I guess it is saver to just keep it in place.

**Review by mchacki — CHANGES_REQUESTED** [review]
Most of it is good!

I only have two suggestions here.

File: arangod/Cluster/Maintenance.cpp (hunk 1 above)
Comment (neunhoef) [reply arangod/Cluster/Maintenance.cpp]: Done.

File: arangod/Cluster/Maintenance.cpp (hunk 2 above)
Comment (neunhoef) [reply arangod/Cluster/Maintenance.cpp]: Done.

**Review by mchacki — APPROVED** [review]
LGTM

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11552/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11553/

**Comment by neunhoef** [comment]
https://jenkins01.arangodb.biz/job/arangodb-matrix-pr-linux-nonmaintainercompile/10928/
(rerun because of Jenkins problems)

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/13208/

**Comment by neunhoef** [comment]
https://jenkins01.arangodb.biz/job/arangodb-matrix-pr/11559/

**Comment by KVS85** [comment]
Tests blue.

**Pull request merged**